psig_matcher/experiments/run_experiment.py

```diff
@@ -120,11 +120,45 @@ def run_meta_markov_multivariant_analysis(parts: List[Part], part_dim: int, num_
         collision_rate = estimate_overlap_of_set_with_sample_signals(multivariant_parts, num_samples, meta_pdf_ci, part_pdf_ci, confidence_bound)
         
         collisions.append(collision_rate)
+        log_metric("monte_carlo_collision_rate", collision_rate)
+
         lower, upper = compute_normal_ci(collisions, 0.95)
         confidence_ranges.append(upper - lower)
+        log_metric("monte_carlo_confidence_interval", upper - lower)
+
         # print(f"Estimated collision rate from sample distributiion has range: {upper - lower}")
         
         if len(confidence_ranges) > 100 and np.mean(confidence_ranges[-10:]) >= np.mean(confidence_ranges[-100:]):
+
+            # final approximated collision rate, and its associated CI
+            log_metric("monte_carlo_final_collision_rate", collision_rate)
+            log_metric("monte_carlo_final_confidence_interval", upper - lower)
+
+            return upper
+
+def simulate_part_pdf_convergence(part_signals: np.ndarray, part_dim: int, part_pdf_ci: float):
+    """ Given a discrete set of signals, this will simulate the part PDF CI convergence methodology.
+    This function pretend that the set of the signals is infinite, Also incorporate logic to handle
+    if we didn't converge before we ran out of data. Have modular connection style such that we could
+    add streaming data source in the future. """
+    sub_samples = []
+    confidence_ranges = []
+    while part_signals:
+        part_signal = part_signals[0]
+        part_signals = part_signals[1:]
```

Comment:
This seems a bit unintuitive to me. For readability, just use pop

Reply:
@henrygilbert22 

I was unable to find a way to do this since pop() does not exist in ndarray.

I have updated this part with detailed comments for readability and clarity.